File: Heap/JAVA/src/implementing_heap/MaxHeap.java
package implementing_heap;

import java.util.ArrayList;

public class MaxHeap {
	public void heapify(ArrayList<Integer> arr, int i) {
		int largest = i;
		int size = arr.size();
		int leftChild = 2 * i + 1;
		int rightChild = 2 * i + 2;

		if (leftChild < size && arr.get(leftChild) > arr.get(largest))
			largest = leftChild;
		if (rightChild < size && arr.get(rightChild) > arr.get(largest))
			largest = rightChild;
		if (largest != i) {
			int temp = arr.get(largest);
			arr.set(largest, arr.get(i));
			arr.set(i, temp);
			heapify(arr, largest);
		}
	}

	public void insert(ArrayList<Integer> arr, int num) {
		int size = arr.size();
		if (size == 0)
			arr.add(num);
		else {
			arr.add(num);
			for (int i = size / 2; i >= 0; i--)
				heapify(arr, i);
		}
	}

	public void delete(ArrayList<Integer> arr, int num) {
		int size = arr.size();
		int i;
		for (i = 0; i < size; i++) {
			if (num == arr.get(i))
				break;
		}
		int temp = arr.get(i);
		arr.set(i, arr.get(size - 1));
		arr.set(size - 1, temp);
		arr.remove(size - 1);
		for (int j = size / 2; j >= 0; j--)
			heapify(arr, j);
	}

	public void peek(ArrayList<Integer> arr) {
		System.out.println(arr.get(0));
	}

	public int extract(ArrayList<Integer> arr) {
		int ret = arr.get(0);
		delete(arr, ret);
		return ret;
	}

	public void printHeap(ArrayList<Integer> arr) {
		for (int each : arr)
			System.out.print(each + "   ");
		System.out.println();
	}

	public static void main(String[] args) {
		MaxHeap myHeap = new MaxHeap();
		ArrayList<Integer> arr = new ArrayList<Integer>();

		for (int i = 1; i < 54; i = i + 3)
			myHeap.insert(arr, i);

		myHeap.printHeap(arr);

		for (int i = 10; i < 20; i += 2) {
			try {
				myHeap.delete(arr, i);
			} catch (IndexOutOfBoundsException e) {
				continue;
			}
		}

		myHeap.printHeap(arr);

		myHeap.peek(arr);
		System.out.println(myHeap.extract(arr));
		myHeap.printHeap(arr);
	}
}

File: Heap/Python/MaxHeap.py
def heapify(arr, size, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2
    if left < size and arr[i] < arr[left]:
        largest = left
    if right < size and arr[i] < arr[right]:
        largest = right
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Swap
        heapify(arr, size, largest)


def insert(arr, num):
    size = len(arr)
    if size == 0:
        arr.append(num)
    else:
        arr.append(num)
        for i in range(size // 2, -1, -1):  # From size//2 to 0 in steps of -1
            heapify(arr, size + 1, i)


def delete(arr, num):
    size = len(arr)
    i = 0
    for i in range(0, size):
        if num == arr[i]:
            break
    arr[i], arr[size - 1] = arr[size - 1], arr[i]  # Swap
    arr.remove(num)
    for i in range(size // 2, -1, -1):
        heapify(arr, size - 1, i)


def peek(arr):
    print(arr[0])


def extract(arr):
    ret = arr[0]
    delete(arr, arr[0])
    return ret


my_heap = []

for j in range(0, 10):
    insert(my_heap, j)
for k in range(1, 5, 2):
    delete(my_heap, k)

print(my_heap)

peek(my_heap)
print(extract(my_heap))
